algebra-kernels/src/fft.rs

```diff
@@ -0,0 +1,191 @@
+use algebra_cl_gen::gpu::{GPUError, GPUResult, kernel_fft};
+use algebra::{
+    PrimeField
+};
+use log::{error};
+use rust_gpu_tools::*;
+
+use lazy_mut::LazyMut;
+use std::any::TypeId;
+use std::cmp;
+use std::env;
+
+// use crate::utils::CACHED_PROGRAMS;
+use std::collections::HashMap;
+
+const LOG2_MAX_ELEMENTS: usize = 32; // At most 2^32 elements is supported.
+const MAX_LOG2_RADIX: u32 = 8; // Radix256
+const MAX_LOG2_LOCAL_WORK_SIZE: u32 = 7; // 128
+
+pub fn get_gpu_min_length() -> usize {
+
+    env::var("FFT_GPU_MIN_LENGTH")
+        .and_then(|v| match v.parse() {
+            Ok(val) => Ok(val),
+            Err(_) => {
+                error!("Invalid FFT_GPU_MIN_LENGTH! Defaulting to 128...");
+                Ok(128)
+            }
+        })
+        .unwrap_or(128)
+}
+
+static mut CACHED_PROGRAMS: LazyMut<HashMap<opencl::Device, HashMap<TypeId, opencl::Program>>> = LazyMut::Init(HashMap::<opencl::Device, HashMap<TypeId, opencl::Program>>::new);
```

Comment:
Why you need the unsafe {} block to access this variable ? Looking at lazy_mut doc it seems to be unnecessary https://docs.rs/lazy_mut/0.1.0/lazy_mut/ , maybe you have to declare the variable using the macro "lazy_mut!". Plus, do we really need this lazy_mut crate ? You can also use a RefCell<HashMap<...>> and wrap it in a lazy_static!() ?

Reply:
Plus, if you use a lazy_static! or lazy_mut! macro, you can call the init() function already in there.
What is the purpose of having this map mutable anyway ? To add new devices at run-time ?

Reply:
I've rewritten the code to avoid the duplication of same parts. But RefCell feature doesn't work for me - I am getting the compilation-time error that using RefCell in this way is not thread-safe. Also without unsafe parameter I am getting compilation-time error "this is unsafe call".

Reply:
If you need to access it from multiple threads at a time, RefCell won't do. You should use Mutex or RWLock.
This article is very useful:
https://manishearth.github.io/blog/2015/05/27/wrapper-types-in-rust-choosing-your-guarantees/

Reply:
I've spent a lot of time with different combinations of these reference patterns, but no success. All variants: Mutex, RWLock - lead me to the situation, when I am trying to pass out of function local variable (opencl::Program), and the opencl::Program doesn't implement Copy trait. I don't have any idea how to not use LazyMut here

Reply:
It's just a minor thing, it's ok